RFCs/0009-Fn::For.md

```diff
@@ -0,0 +1,903 @@
+# Motivation
+
+In a CloudFormation template, a single Resource configures one infrastructure object. Therefore, a template can become verbose when you manually declare multiple similar resources. Some examples would include a pool of EC2 instances with the same configurations but different instance types, or S3 Bucket Notification configurations with different SNS Topics.
+
+# Declaration
+## Fn::Map
+### Json
+```json
+{
+  "Fn::Map": {
+    "index": "Index",
+    "value": "Value",
+    "collection": "CollectionToIterate",
+    "fragment": { },
+    "key": "Key"
+  }
+}
+```
+
+### Yaml
+```yaml
+Fn::Map: # or !Map
+  index: "Index"
+  value: "Value"
+  collection: "CollectionToIterate"
+  fragment: { }
+  key: "Key"
+```
+
+### Parameters
+
+Fn::Map supports 5 parameters.
+
+* `index` (optional)
+    * The name of an iterator variable which stores a positional index of each value in a given List.
+    * The index number sequence starts from 0.
+    * String type
+    * the default value is "index"
+* `value` (optional)
+    * The name of an iterator variable which stores each value in a List.
+    * String type
+    * the default value is "value"
+* `collection`
+    * The collection to iterate on. It must be a List. You can either in-place the collection itself or reference List type parameter values from the Parameters section.
+* `fragment`
+    * The fragment to replicate for each item in iteration
+* `key` (conditional)
+    * The json or yaml key for replicated objects. For Resources and Outputs this will be the logical id.
+    * **This field is only required when Fn::Map should generate an object of key value pairs, for example in the Resources or Outputs section of a template. It should not be specified when using Fn::Map to generate an array, for example a list type attribute for a Resource Property.** 
+    * The value **should be interpolated with an index** using `Fn::Sub` intrinsic function.
+
+
+## Fn::Merge
+Fn::Merge takes an array of objects as input and merges them into the parent object
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": { },
+      "key2": { }
+    },
+    {
+      "key3": { }
+    }
+  ],
+  "key4": { }
+}
+```
+
+Result:
+```json
+{
+  "key1": { },
+  "key2": { },
+  "key3": { },
+  "key4": { }
+}
+```
+
+Fn::Merge will throw an error when there is a key collision. The below examples will cause an error:
+
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": {}
+    },
+    {
+      "key1": { }
+    }
+  ]
+}
+```
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": {}
+    }
+  ],
+  "key1": { }
+}
+```
+
+# Examples
+### Usage in Resources Section
+#### Use case: Replicate a single Resource
+* Note that the **`key`** parameter is provided.
+
+In Json:
+```json
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Description": "EC2 Instances with different AMIs",
+  "Resources": {
+    "Fn::Merge": [
```

Comment:
Could CloudFormation be smart enough to know it should Merge if there is a `key` parameter?